File: src/backend-contract/background-continuation.ts
// src/backend-contract/background-continuation.ts
//
// `background.continuation`: the declared standing order the OS wake executes
// (BGS4, the 5.0 headline capability). The app declares while alive what a
// wake may do; the wake executes only what was declared — the library never
// invents a connect. `record-only` is today's behaviour and the default, so
// 5.0 changes nothing for an app that does not opt in.
//
// Strategies: `record-only` (install owner, record restored peer, stop),
// `native` (reconnect the declared known peer + resubscribe the declared
// characteristics through the Rust core, no JavaScript), `headless-task` (run
// the registered headless JS task, Android), `foreground-service` (start the
// configured connected-device foreground service from the wake). The deferred
// strategies keep their validated option shape here so rc.1 adds executors
// with no breaking change; until then they answer `capability.unsupported`
// with "not implemented in this release" — never "the platform cannot".

import { contractError } from './errors'

/** Wake strategies in declaration order. `record-only` is the default. */
export const CONTINUATION_STRATEGIES = Object.freeze([
  'record-only',
  'native',
  'headless-task',
  'foreground-service'
] as const)

export type BackgroundContinuationStrategy = (typeof CONTINUATION_STRATEGIES)[number]

/** One capability per strategy, reported at runtime by the instantiated backend. */
export const BACKGROUND_CONTINUATION_FEATURE_IDS = Object.freeze({
  wakeOnAppearance: 'background:wake-on-appearance',
  nativeResubscribe: 'background:native-resubscribe',
  headlessTask: 'background:headless-task',
  wakeNotification: 'background:wake-notification'
} as const)

export type BackgroundContinuationFeatureId =
  (typeof BACKGROUND_CONTINUATION_FEATURE_IDS)[keyof typeof BACKGROUND_CONTINUATION_FEATURE_IDS]

/**
 * Wake outcome vocabulary. ONE vocabulary on both hosts: the wake reports
 * `continuation.completed` or `continuation.failed` with the platform's own
 * detail underneath — the same event names and words on Android and iOS.
 * iOS parity arrives in rc.1; these names are designed so the Apple path
 * reports the same events.
 */
export const CONTINUATION_OUTCOME_EVENTS = Object.freeze(['continuation.completed', 'continuation.failed'] as const)

export type BackgroundContinuationOutcomeEvent = (typeof CONTINUATION_OUTCOME_EVENTS)[number]

/** One declared GATT resubscription for the `native` standing order. */
export interface BackgroundContinuationResubscribeSelector {
  readonly serviceUuid: string
  readonly serviceOccurrence: number
  readonly characteristicUuid: string
  readonly characteristicOccurrence: number
}

/** Foreground-service configuration for the deferred `foreground-service` strategy. */
export interface BackgroundContinuationForegroundService {
  readonly notification: {
    readonly channelId: string
    readonly channelName: string
    readonly title: string
    readonly body?: string
    readonly icon?: string
  }
}

/** The declared standing order: what one OS wake may do, and to what. */
export interface BackgroundContinuationDeclaration {
  readonly onAppearance: BackgroundContinuationStrategy
  /** Uppercase MAC subject; absent scopes the order to whichever armed peer appears. */
  readonly peerId?: string
  readonly resubscribe: readonly BackgroundContinuationResubscribeSelector[]
  readonly headlessTaskName?: string
  readonly foregroundService?: BackgroundContinuationForegroundService
}

/** The default: today's behaviour — install owner, record restored peer, stop. */
export const DEFAULT_BACKGROUND_CONTINUATION: BackgroundContinuationDeclaration = Object.freeze({
  onAppearance: 'record-only' as const,
  resubscribe: Object.freeze([])
})

/**
 * Wake consumer name prefix. The native wake subscribes as
 * `ubm-continuation-{index}` (one per declared resubscription, in order), so
 * the claim maps each backlog value to the selector that subscribed it. The
 * Kotlin executor's `CONSUMER_PREFIX` must stay identical; both sides cite
 * this constant's spelling.
 */
export const CONTINUATION_CONSUMER_PREFIX = 'ubm-continuation-' as const

const DECLARATION_KEYS = Object.freeze([
  'onAppearance',
  'peerId',
  'resubscribe',
  'headlessTaskName',
  'foregroundService'
])
const SELECTOR_KEYS = Object.freeze([
  'serviceUuid',
  'serviceOccurrence',
  'characteristicUuid',
  'characteristicOccurrence'
])
const FOREGROUND_SERVICE_KEYS = Object.freeze(['notification'])
const NOTIFICATION_KEYS = Object.freeze(['channelId', 'channelName', 'title', 'body', 'icon'])

const MAC_PATTERN = /^([0-9A-Fa-f]{2}:){5}[0-9A-Fa-f]{2}$/
const UUID_PATTERN = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/

function isPlainRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function rejectUnknownKeys(value: Record<string, unknown>, expected: readonly string[], label: string): void {
  const unknown = Object.keys(value).filter(key => !expected.includes(key))
  if (unknown.length > 0) {
    throw contractError('argument.invalid', 'restoration', `${label}.unknown-key`)
  }
}

function nonEmptyString(value: unknown, label: string): string {
  if (typeof value !== 'string' || value.length === 0) {
    throw contractError('argument.invalid', 'restoration', label)
  }
  return value
}

function occurrence(value: unknown, label: string): number {
  if (value === undefined) return 1
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 1) {
    throw contractError('argument.invalid', 'restoration', label)
  }
  return value
}

function canonicalUuid(value: unknown, label: string): string {
  const text = nonEmptyString(value, label)
  if (!UUID_PATTERN.test(text)) {
    throw contractError('argument.invalid', 'restoration', label)
  }
  return text.toLowerCase()
}

function selector(value: unknown): BackgroundContinuationResubscribeSelector {
  if (!isPlainRecord(value)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.resubscribe.entry')
  }
  rejectUnknownKeys(value, SELECTOR_KEYS, 'background.continuation.resubscribe.entry')
  return Object.freeze({
    serviceUuid: canonicalUuid(value.serviceUuid, 'background.continuation.resubscribe.serviceUuid'),
    serviceOccurrence: occurrence(value.serviceOccurrence, 'background.continuation.resubscribe.serviceOccurrence'),
    characteristicUuid: canonicalUuid(
      value.characteristicUuid,
      'background.continuation.resubscribe.characteristicUuid'
    ),
    characteristicOccurrence: occurrence(
      value.characteristicOccurrence,
      'background.continuation.resubscribe.characteristicOccurrence'
    )
  })
}

function notification(value: unknown): BackgroundContinuationForegroundService['notification'] {
  if (!isPlainRecord(value)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.foregroundService.notification')
  }
  rejectUnknownKeys(value, NOTIFICATION_KEYS, 'background.continuation.foregroundService.notification')
  const channelId = nonEmptyString(value.channelId, 'background.continuation.foregroundService.notification.channelId')
  const channelName = nonEmptyString(
    value.channelName,
    'background.continuation.foregroundService.notification.channelName'
  )
  const title = nonEmptyString(value.title, 'background.continuation.foregroundService.notification.title')
  const body =
    value.body === undefined
      ? undefined
      : nonEmptyString(value.body, 'background.continuation.foregroundService.notification.body')
  const icon =
    value.icon === undefined
      ? undefined
      : nonEmptyString(value.icon, 'background.continuation.foregroundService.notification.icon')
  return Object.freeze({
    channelId,
    channelName,
    title,
    ...(body === undefined ? {} : { body }),
    ...(icon === undefined ? {} : { icon })
  })
}

/**
 * Normalizes a caller-authored continuation declaration. Unknown strategies
 * and unknown keys are refused — never replaced with a quieter path.
 */
export function normalizeBackgroundContinuation(input: unknown): BackgroundContinuationDeclaration {
  if (input === undefined) return DEFAULT_BACKGROUND_CONTINUATION
  if (!isPlainRecord(input)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation')
  }
  rejectUnknownKeys(input, DECLARATION_KEYS, 'background.continuation')
  const onAppearance = input.onAppearance === undefined ? 'record-only' : input.onAppearance
  if (typeof onAppearance !== 'string' || !(CONTINUATION_STRATEGIES as readonly string[]).includes(onAppearance)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.onAppearance')
  }
  const strategy = onAppearance as BackgroundContinuationStrategy
  const peerId = input.peerId === undefined ? undefined : normalizePeerAddress(input.peerId)
  const resubscribe =
    input.resubscribe === undefined ? Object.freeze([]) : Object.freeze(normalizeResubscribe(input.resubscribe))
  const headlessTaskName =
    input.headlessTaskName === undefined
      ? undefined
      : nonEmptyString(input.headlessTaskName, 'background.continuation.headlessTaskName')
  if (strategy === 'headless-task' && headlessTaskName === undefined) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.headlessTaskName.required')
  }
  if (strategy !== 'headless-task' && headlessTaskName !== undefined) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.headlessTaskName.strategy')
  }
  const foregroundService =
    input.foregroundService === undefined ? undefined : normalizeForegroundService(input.foregroundService)
  if (strategy === 'foreground-service' && foregroundService === undefined) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.foregroundService.required')
  }
  if (strategy !== 'foreground-service' && foregroundService !== undefined) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.foregroundService.strategy')
  }
  return Object.freeze({
    onAppearance: strategy,
    ...(peerId === undefined ? {} : { peerId }),
    resubscribe,
    ...(headlessTaskName === undefined ? {} : { headlessTaskName }),
    ...(foregroundService === undefined ? {} : { foregroundService })
  })
}

function normalizePeerAddress(value: unknown): string {
  const text = nonEmptyString(value, 'background.continuation.peerId')
  if (!MAC_PATTERN.test(text)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.peerId')
  }
  return text.toUpperCase()
}

function normalizeResubscribe(value: unknown): BackgroundContinuationResubscribeSelector[] {
  if (!Array.isArray(value)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.resubscribe')
  }
  if (value.length > 64) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.resubscribe.too-many')
  }
  return value.map(selector)
}

/**
 * Canonical JSON of a normalized declaration for native persistence (the
 * shape `BackgroundContinuationStore` parses). One authority: the binding
 * serializes here, never hand-builds the payload.
 */
export function serializeBackgroundContinuation(declaration: BackgroundContinuationDeclaration): string {
  return JSON.stringify({
    onAppearance: declaration.onAppearance,
    ...(declaration.peerId === undefined ? {} : { peerId: declaration.peerId }),
    resubscribe: declaration.resubscribe.map(entry => ({
      serviceUuid: entry.serviceUuid,
      serviceOccurrence: entry.serviceOccurrence,
      characteristicUuid: entry.characteristicUuid,
      characteristicOccurrence: entry.characteristicOccurrence
    })),
    ...(declaration.headlessTaskName === undefined ? {} : { headlessTaskName: declaration.headlessTaskName }),
    ...(declaration.foregroundService === undefined ? {} : { foregroundService: declaration.foregroundService })
  })
}

function normalizeForegroundService(value: unknown): BackgroundContinuationForegroundService {
  if (!isPlainRecord(value)) {
    throw contractError('argument.invalid', 'restoration', 'background.continuation.foregroundService')
  }
  rejectUnknownKeys(value, FOREGROUND_SERVICE_KEYS, 'background.continuation.foregroundService')
  if (value.notification === undefined) {
    throw contractError(
      'argument.invalid',
      'restoration',
      'background.continuation.foregroundService.notification.required'
    )
  }
  return Object.freeze({ notification: notification(value.notification) })
}

File: android/src/main/java/com/sfourdrinier/unifiedblemanager/presence/BackgroundContinuation.kt
// android/src/main/java/com/sfourdrinier/unifiedblemanager/presence/BackgroundContinuation.kt

package com.sfourdrinier.unifiedblemanager.presence

import com.sfourdrinier.unifiedblemanager.rustcore.RustCoreJson

/** Wake strategies of the declared standing order, in declaration order. */
enum class ContinuationStrategy(val wire: String) {
  RECORD_ONLY("record-only"),
  NATIVE("native"),
  HEADLESS_TASK("headless-task"),
  FOREGROUND_SERVICE("foreground-service");

  companion object {
    fun parse(wire: Any?, operation: String): ContinuationStrategy {
      val text = wire as? String
        ?: throw IllegalArgumentException("$operation: onAppearance must be a string")
      return values().firstOrNull { it.wire == text }
        ?: throw IllegalArgumentException("$operation: unknown onAppearance $text")
    }
  }
}

/** One declared GATT resubscription of the `native` standing order. */
data class ContinuationSelector(
  val serviceUuid: String,
  val serviceOccurrence: Long,
  val characteristicUuid: String,
  val characteristicOccurrence: Long
)

/** Foreground-service configuration of the deferred `foreground-service` strategy. */
data class ContinuationForegroundService(val notification: ContinuationNotification)

/** Notification of the deferred `foreground-service` strategy. */
data class ContinuationNotification(
  val channelId: String,
  val channelName: String,
  val title: String,
  val body: String?,
  val icon: String?
)

/**
 * The declared standing order the OS wake executes (BGS4). Parsed from the
 * canonical JSON the binding persists — the same key set
 * `serializeBackgroundContinuation` writes, so a payload the JavaScript side
 * could not produce is refused here, never executed.
 */
data class BackgroundContinuationDeclaration(
  val strategy: ContinuationStrategy,
  /** Uppercase MAC subject; null scopes the order to whichever armed peer appears. */
  val peerId: String?,
  val resubscribe: List<ContinuationSelector>,
  val headlessTaskName: String?,
  val foregroundService: ContinuationForegroundService?
) {
  companion object {
    private const val OPERATION = "background.continuation"
    private val TOP_KEYS = setOf("onAppearance", "peerId", "resubscribe", "headlessTaskName", "foregroundService")
    private val SELECTOR_KEYS = setOf("serviceUuid", "serviceOccurrence", "characteristicUuid", "characteristicOccurrence")
    private val MAC = Regex("^([0-9A-Fa-f]{2}:){5}[0-9A-Fa-f]{2}$")
    private val UUID = Regex("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")

    fun recordOnly(): BackgroundContinuationDeclaration = BackgroundContinuationDeclaration(
      strategy = ContinuationStrategy.RECORD_ONLY,
      peerId = null,
      resubscribe = emptyList(),
      headlessTaskName = null,
      foregroundService = null
    )

    /** Parses persisted JSON; null (never declared) means `record-only`. */
    fun parse(json: String?): BackgroundContinuationDeclaration {
      if (json == null) return recordOnly()
      val root = try {
        RustCoreJson.parse(json) as? Map<*, *> ?: throw IllegalArgumentException("$OPERATION: not an object")
      } catch (error: RustCoreJson.MalformedJson) {
        throw IllegalArgumentException("$OPERATION: malformed: ${error.message}")
      }
      rejectUnknown(root.keys, TOP_KEYS, OPERATION)
      val strategy = if (!root.containsKey("onAppearance")) ContinuationStrategy.RECORD_ONLY
      else ContinuationStrategy.parse(root["onAppearance"], OPERATION)
      val peerId = if (!root.containsKey("peerId")) null else {
        val text = root["peerId"] as? String
          ?: throw IllegalArgumentException("$OPERATION: peerId must be a string")
        if (!MAC.matches(text)) throw IllegalArgumentException("$OPERATION: peerId must be a MAC address")
        text.uppercase()
      }
      val resubscribe = if (!root.containsKey("resubscribe")) emptyList()
      else {
        val entries = root["resubscribe"] as? List<*>
          ?: throw IllegalArgumentException("$OPERATION: resubscribe must be an array")
        if (entries.size > 64) throw IllegalArgumentException("$OPERATION: resubscribe too many")
        entries.map { selector(it) }
      }
      val headlessTaskName = if (!root.containsKey("headlessTaskName")) null else {
        val text = root["headlessTaskName"] as? String
          ?: throw IllegalArgumentException("$OPERATION: headlessTaskName must be a string")
        if (text.isEmpty()) throw IllegalArgumentException("$OPERATION: headlessTaskName must be non-empty")
        text
      }
      if (strategy == ContinuationStrategy.HEADLESS_TASK && headlessTaskName == null) {
        throw IllegalArgumentException("$OPERATION: headlessTaskName required for headless-task")
      }
      if (strategy != ContinuationStrategy.HEADLESS_TASK && headlessTaskName != null) {
        throw IllegalArgumentException("$OPERATION: headlessTaskName applies only to headless-task")
      }
      val foregroundService = if (!root.containsKey("foregroundService")) null
      else foregroundService(root["foregroundService"])
      if (strategy == ContinuationStrategy.FOREGROUND_SERVICE && foregroundService == null) {
        throw IllegalArgumentException("$OPERATION: foregroundService required for foreground-service")
      }
      if (strategy != ContinuationStrategy.FOREGROUND_SERVICE && foregroundService != null) {
        throw IllegalArgumentException("$OPERATION: foregroundService applies only to foreground-service")
      }
      return BackgroundContinuationDeclaration(strategy, peerId, resubscribe, headlessTaskName, foregroundService)
    }

    private fun selector(value: Any?): ContinuationSelector {
      val entry = value as? Map<*, *>
        ?: throw IllegalArgumentException("$OPERATION: resubscribe entry must be an object")
      rejectUnknown(entry.keys, SELECTOR_KEYS, "$OPERATION resubscribe entry")
      return ContinuationSelector(
        serviceUuid = uuid(entry["serviceUuid"], "$OPERATION resubscribe serviceUuid"),
        serviceOccurrence = occurrence(entry["serviceOccurrence"]),
        characteristicUuid = uuid(entry["characteristicUuid"], "$OPERATION resubscribe characteristicUuid"),
        characteristicOccurrence = occurrence(entry["characteristicOccurrence"])
      )
    }

    private fun uuid(value: Any?, label: String): String {
      val text = value as? String ?: throw IllegalArgumentException("$label must be a string")
      if (!UUID.matches(text)) throw IllegalArgumentException("$label must be a canonical 128-bit UUID")
      return text.lowercase()
    }

    private fun occurrence(value: Any?): Long {
      if (value == null) return 1L
      val number = (value as? Number)?.toLong()
        ?: throw IllegalArgumentException("$OPERATION resubscribe occurrence must be a positive integer")
      if (number < 1L) throw IllegalArgumentException("$OPERATION resubscribe occurrence must be a positive integer")
      return number
    }

    private fun foregroundService(value: Any?): ContinuationForegroundService {
      val service = value as? Map<*, *>
        ?: throw IllegalArgumentException("$OPERATION: foregroundService must be an object")
      rejectUnknown(service.keys, setOf("notification"), "$OPERATION foregroundService")
      val raw = service["notification"] as? Map<*, *>
        ?: throw IllegalArgumentException("$OPERATION: foregroundService notification required")
      rejectUnknown(raw.keys, setOf("channelId", "channelName", "title", "body", "icon"), "$OPERATION notification")
      fun text(key: String): String {
        val text = raw[key] as? String
          ?: throw IllegalArgumentException("$OPERATION: notification $key must be a string")
        if (text.isEmpty()) throw IllegalArgumentException("$OPERATION: notification $key must be non-empty")
        return text
      }
      return ContinuationForegroundService(
        ContinuationNotification(
          channelId = text("channelId"),
          channelName = text("channelName"),
          title = text("title"),
          body = raw["body"] as? String,
          icon = raw["icon"] as? String
        )
      )
    }

    private fun rejectUnknown(keys: Set<Any?>, expected: Set<String>, label: String) {
      val unknown = keys.filterIsInstance<String>().filter { it !in expected }.sorted()
      if (unknown.isNotEmpty()) throw IllegalArgumentException("$label unknown keys: ${unknown.joinToString(",")}")
    }
  }
}

/**
 * The wake outcome in ONE vocabulary on both hosts: `continuation.completed`
 * or `continuation.failed` with the platform's own reason underneath — the
 * same event names and words on Android and iOS.
 */
sealed interface ContinuationOutcome {
  val strategy: ContinuationStrategy
  val event: String

  data class Completed(
    override val strategy: ContinuationStrategy,
    val peerAddress: String,
    val resubscribed: Int
  ) : ContinuationOutcome {
    override val event = "continuation.completed"
  }

  data class Failed(
    override val strategy: ContinuationStrategy,
    /** Public error code (`capability.unsupported`, `connection.failed`, …). */
    val code: String,
    /** Human reason; the platform's own detail rides in [platform]. */
    val reason: String,
    /** The platform's own error identity, or null when the failure is ours. */
    val platform: String?
  ) : ContinuationOutcome {
    override val event = "continuation.failed"
  }

  companion object {
    fun completed(strategy: ContinuationStrategy, peerAddress: String, resubscribed: Int): ContinuationOutcome =
      Completed(strategy, peerAddress, resubscribed)

    fun failed(strategy: ContinuationStrategy, code: String, reason: String, platform: String?): ContinuationOutcome =
      Failed(strategy, code, reason, platform)
  }
}

/** The last-wake record Diagnostics renders (and unsupported states quote). */
data class ContinuationWakeRecord(
  val observedAtMs: Long,
  val event: String,
  val strategy: ContinuationStrategy,
  val peerAddress: String?,
  val code: String?,
  val reason: String?
)
